src/historywork/WriteVerifiedCheckpointHashesWork.cpp

```diff
@@ -0,0 +1,169 @@
+// Copyright 2020 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "WriteVerifiedCheckpointHashesWork.h"
+#include "catchup/CatchupConfiguration.h"
+#include "catchup/CatchupRange.h"
+#include "catchup/DownloadApplyTxsWork.h"
+#include "catchup/VerifyLedgerChainWork.h"
+#include "history/FileTransferInfo.h"
+#include "history/HistoryManager.h"
+#include "historywork/BatchDownloadWork.h"
+#include "ledger/LedgerManager.h"
+#include "ledger/LedgerRange.h"
+#include "main/Application.h"
+#include "util/Logging.h"
+#include "work/ConditionalWork.h"
+#include <Tracy.hpp>
+#include <fmt/format.h>
+
+namespace stellar
+{
+
+class TmpDirOwnerWork : public Work
+{
+    TmpDir const mTmpDir;
+
+  public:
+    TmpDirOwnerWork(Application& app, std::string const& name)
+        : Work(app, name, RETRY_NEVER)
+        , mTmpDir(mApp.getTmpDirManager().tmpDir(getName()))
+    {
+    }
+    BasicWork::State
+    doWork() override
+    {
+        return State::WORK_SUCCESS;
+    }
+    TmpDir const&
+    getTmpDir()
+    {
+        return mTmpDir;
+    }
+};
+
+WriteVerifiedCheckpointHashesWork::WriteVerifiedCheckpointHashesWork(
+    Application& app, LedgerNumHashPair rangeEnd, std::string const& outputFile,
+    std::shared_ptr<HistoryArchive> archive)
+    : BatchWork(app, "write-verified-checkpoint-hashes")
+    , mRangeEnd(rangeEnd)
+    , mCurrCheckpoint(rangeEnd.first)
+    , mArchive(archive)
+    , mOutputFile(std::make_shared<std::ofstream>(outputFile))
+{
+    if (mArchive)
+    {
+        CLOG(INFO, "History")
+            << "WriteVerifiedCheckpointHashesWork: selected archive "
+            << mArchive->getName();
+    }
+    if (!*mOutputFile)
+    {
+        throw std::runtime_error("error opening output file " + outputFile);
+    }
+    (*mOutputFile) << "[\n";
+}
+
+WriteVerifiedCheckpointHashesWork::~WriteVerifiedCheckpointHashesWork()
+{
+    (*mOutputFile) << "\n]\n";
+}
+
+bool
+WriteVerifiedCheckpointHashesWork::hasNext() const
+{
+    return mCurrCheckpoint != 0;
+}
+
+std::shared_ptr<BasicWork>
+WriteVerifiedCheckpointHashesWork::yieldMoreWork()
+{
+    if (!hasNext())
+    {
+        throw std::runtime_error("Nothing to iterate over!");
+    }
+
+    auto& hm = mApp.getHistoryManager();
+    uint32_t freq = hm.getCheckpointFrequency();
+
+    auto lclHe = mApp.getLedgerManager().getLastClosedLedgerHeader();
+    LedgerNumHashPair lcl(lclHe.header.ledgerSeq,
+                          make_optional<Hash>(lclHe.hash));
+
+    // We form an _inner_ BatchWork that downloads a bunch of checkpoints in
```

Comment:
Comment suggestions: I think right now the comment gives too many implementation details of other classes (using `curl`, assumptions about BatchDownloadWork deriving from BatchWork etc). I think really what we're trying to fix here is "make sure works that depend on each other don't stall works that can execute in parallel too much". So the solution is to increase "unit of verification" so reduce number of dependencies. Something generic like that. I also don't mind if you want to leave it as is, just seems like the comment gives away a bit too much right now.

Reply:
Fixed (also moved residual comment to header where constant is declared)